src/circuit/circuit_frame.rs

```diff
@@ -4762,9 +4845,9 @@ mod tests {
             assert!(delta == Delta::Equal);
 
             //println!("{}", print_cs(&cs));
-            assert_eq!(12529, cs.num_constraints());
+            assert_eq!(11569, cs.num_constraints());
             assert_eq!(13, cs.num_inputs());
-            assert_eq!(12183, cs.aux().len());
+            assert_eq!(11187, cs.aux().len());
```

Comment:
Another 1K constraints down, with improved correctness and soundness. Maybe 10K is possible, but the lowest-hanging fruit has been picked.